scalding-core/src/main/scala/com/twitter/scalding/Stats.scala

```diff
@@ -48,24 +52,76 @@ object StatKey {
 }
 
 private[scalding] object CounterImpl {
-  def apply(fp: FlowProcess[_], statKey: StatKey): CounterImpl =
-    fp match {
-      case hFP: HadoopFlowProcess => HadoopFlowPCounterImpl(hFP, statKey)
-      case _ => GenericFlowPCounterImpl(fp, statKey)
+  val CounterImplClass = "scalding.stats.counter.impl.class"
+
+  /*
+  Note: knowing which counter implementation class to use is a fabric-dependent question. Prior to Cascading 3, the
+  question of "which fabric" was easy: it was either Local or Hadoop, and both types were available wherever Scalding
+  is loaded.
+
+  We cannot have access to the fabric's Mode / ExecutionMode implementation, because instantiating counters happens
+  at clusterside runtime, not planning time[*]. The FlowProcess[_] detailed type typically also depends on the fabric,
+  with exceptions: Legacy Hadoop (1.x) and Hadoop2-MR1 use the same FlowProcess implementation class name,
+  HadoopFlowProcess.
+  Anyway, we can't use pattern-matching on the specific FlowProcess[_] type, as this would require scalding-core to
+  know about specific fabric implementation details, which it can no longer do (post-Cascading 3).
+
+  Solution found for now:
+    1. the fabric-specific ExecutionMode is responsible for setting the class name of the appropriate counter
+    implementation in a config key (named CounterImplClass above)
+    2. clusterside, CounterImpl$ retrieves this class name and uses reflection to instanciate the correct counter
+    implementation.
+
+
+  [*] even if Mode was accessible at runtime, it'd require a pervasive addition of (possibly implicit) parameters
+    whenever a join between pipes is possible, which would cause significant changes to the API
+
+   */
+  def apply(fp: FlowProcess[_], statKey: StatKey): CounterImpl = {
+    /* TODO: comment this stuff. */
+    val klassName = Option(fp.getStringProperty(CounterImplClass)).getOrElse(sys.error(CounterImplClass + " property is missing"))
+
+    val memo = scala.collection.mutable.Map[String, Constructor[CounterImpl]]()
+    val ctor = memo.synchronized {
+      memo.getOrElse(klassName,
+        {
+          val klass = Class.forName(klassName).asInstanceOf[Class[CounterImpl]]
+          val ctor = klass.getConstructor(classOf[FlowProcess[_]], classOf[StatKey])
+          memo.put(klassName, ctor)
+          ctor
+        })
     }
+    ctor.newInstance(fp, statKey)
+  }
+
+  private[scalding] def upcast[T <: FlowProcess[_]](fp: FlowProcess[_])(implicit ev: TypeTag[T]): T = fp match {
+    case hfp: T @unchecked if (ev == typeTag[T]) => hfp // see below
```

Comment:
this `typeTag[T]` will always match `ev` no? I think so. Why not just explicitly cast in this method:
```scala
def downCast[T <: U](u: U): T = u.asInstanceOf[T]
```
Also, this is a downcast, no?